From 807ce47b82d35f1ce7f425d2b988162fb11617d2 Mon Sep 17 00:00:00 2001
From: Brian Bouterse <bmbouter@gmail.com>
Date: Fri, 19 Dec 2014 14:10:38 -0500
Subject: [PATCH] fixes celery/kombu#439 better handling for a purge of missing
 Queue

---
 kombu/tests/transport/test_qpid.py | 5 +++++
 kombu/transport/qpid.py            | 5 ++++-
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/kombu/tests/transport/test_qpid.py b/kombu/tests/transport/test_qpid.py
index eb7d8f5..390e09c 100644
--- a/kombu/tests/transport/test_qpid.py
+++ b/kombu/tests/transport/test_qpid.py
@@ -522,6 +522,11 @@ class TestChannelPurge(ChannelTestBase):
         result = self.channel._purge(self.mock_queue)
         self.assertEqual(result, 5)
 
+    @patch(QPID_MODULE + '.ChannelError', new=MockException)
+    def test_channel__purge_raises_channel_error_if_queue_does_not_exist(self):
+        self.mock_broker_agent.return_value.getQueue.return_value = None
+        self.assertRaises(MockException, self.channel._purge, self.mock_queue)
+
 
 @case_no_python3
 @case_no_pypy
diff --git a/kombu/transport/qpid.py b/kombu/transport/qpid.py
index ef88bc5..474fe73 100644
--- a/kombu/transport/qpid.py
+++ b/kombu/transport/qpid.py
@@ -62,7 +62,7 @@ try:
 except ImportError:  # pragma: no cover
     qpid = None
 
-
+from kombu.exceptions import ChannelError
 from kombu.five import Empty, items
 from kombu.log import get_logger
 from kombu.transport.virtual import Base64, Message
@@ -503,6 +503,9 @@ class Channel(base.StdChannel):
         :rtype: int
         """
         queue_to_purge = self._broker.getQueue(queue)
+        if queue_to_purge is None:
+            raise ChannelError("queue.purge: server channel error 404, message: "
+                               "NOT_FOUND - no queue '%s'" % queue)
         message_count = queue_to_purge.values['msgDepth']
         if message_count > 0:
             queue_to_purge.purge(message_count)
-- 
1.9.3

